File: spx-backend/internal/model/asset.go
package model

import "fmt"

// Asset is the model for assets.
type Asset struct {
	Model

	// OwnerID is the ID of the asset owner.
	OwnerID int64 `gorm:"column:owner_id;index"`
	Owner   User  `gorm:"foreignKey:OwnerID"`

	// DisplayName is the display name.
	DisplayName string `gorm:"column:display_name;index:,class:FULLTEXT"`

	// Type is the type of the asset.
	Type AssetType `gorm:"column:type;index"`

	// Category is the category to which the asset belongs.
	Category string `gorm:"column:category;index"`

	// Files contains the file paths and their corresponding universal URLs
	// associated with the asset.
	Files FileCollection `gorm:"column:files"`

	// FilesHash is the hash of the asset files.
	FilesHash string `gorm:"column:files_hash"`

	// Visibility is the visibility.
	Visibility Visibility `gorm:"column:visibility;index"`
}

// TableName implements [gorm.io/gorm/schema.Tabler].
func (Asset) TableName() string {
	return "asset"
}

// AssetType is the type of asset.
type AssetType uint8

const (
	AssetTypeSprite AssetType = iota
	AssetTypeBackdrop
	AssetTypeSound
)

// ParseAssetType parses the string representation of an asset type.
func ParseAssetType(s string) (AssetType, error) {
	var at AssetType
	return at, at.UnmarshalText([]byte(s))
}

// String implements [fmt.Stringer].
func (at AssetType) String() string {
	if text, err := at.MarshalText(); err == nil {
		return string(text)
	}
	return fmt.Sprintf("AssetType(%d)", at)
}

// MarshalText implements [encoding.TextMarshaler].
func (at AssetType) MarshalText() ([]byte, error) {
	switch at {
	case AssetTypeSprite:
		return []byte("sprite"), nil
	case AssetTypeBackdrop:
		return []byte("backdrop"), nil
	case AssetTypeSound:
		return []byte("sound"), nil
	default:
		return nil, fmt.Errorf("invalid asset type: %d", at)
	}
}

// UnmarshalText implements [encoding.TextUnmarshaler].
func (at *AssetType) UnmarshalText(text []byte) error {
	switch string(text) {
	case "sprite":
		*at = AssetTypeSprite
	case "backdrop":
		*at = AssetTypeBackdrop
	case "sound":
		*at = AssetTypeSound
	default:
		return fmt.Errorf("invalid asset type: %s", text)
	}
	return nil
}

File: spx-gui/src/apis/asset.ts
import type { Prettify } from '@/utils/types'
import type {
  FileCollection,
  ByPage,
  PaginationParams,
  Perspective,
  ArtStyle,
  SpriteCategory,
  BackdropCategory,
  SoundCategory
} from './common'
import { client, Visibility } from './common'

export { Visibility }

export const assetDisplayNameMaxLength = 100

export enum AssetType {
  Sprite = 'sprite',
  Backdrop = 'backdrop',
  Sound = 'sound'
}

export type AssetExtraSettings = {
  /** Category to which the asset belongs */
  category?: SpriteCategory | BackdropCategory | SoundCategory
  /**
   * Art style indicates the visual style or aesthetic approach used in the creation of graphics.
   * NOTE: Not available for Sound assets.
   */
  artStyle?: ArtStyle
  /**
   * Perspective indicates the viewpoint from which the "game world" is viewed.
   * NOTE: Not available for Sound assets.
   */
  perspective?: Perspective
}

export type AssetData = {
  /** Unique identifier */
  id: string
  /** Username of the asset's owner */
  owner: string
  /** Display name of the asset */
  displayName: string
  /** Type of the asset */
  type: AssetType
  /**
   * Category to which the asset belongs
   * @deprecated Use `category` in `extraSettings` instead.
   */
  category: string
  /** Brief description of the asset */
  description: string
  /** Extra settings specific to the asset */
  extraSettings: AssetExtraSettings
  /** File paths and their corresponding universal URLs associated with the asset */
  files: FileCollection
  /** Hash of the asset files */
  filesHash: string
  /** Visibility of the asset */
  visibility: Visibility
}

export type ListAssetsParams = PaginationParams & {
  /** Filter assets by display name pattern */
  keyword?: string
  /** Filter assets by type */
  type?: AssetType
  /**
   * Filter assets by category
   * @deprecated Not recommended for use as the field `category` is deprecated.
   */
  category?: string
  /** Filter assets by files hash */
  filesHash?: string
  /** Filter assets by visibility */
  visibility?: Visibility
  /** Field by which to order the results */
  orderBy?: 'createdAt' | 'updatedAt' | 'displayName'
  /** Order in which to sort the results */
  sortOrder?: 'asc' | 'desc'
}

export function listAssets(params?: ListAssetsParams, signal?: AbortSignal) {
  return client.get('/assets', params, { signal }) as Promise<ByPage<AssetData>>
}

export function listSignedInUserAssets(params?: ListAssetsParams, signal?: AbortSignal) {
  return client.get('/user/assets', params, { signal }) as Promise<ByPage<AssetData>>
}

export type AddAssetParams = Prettify<
  Pick<
    AssetData,
    'displayName' | 'type' | 'category' | 'description' | 'extraSettings' | 'files' | 'filesHash' | 'visibility'
  >
>

export function addAsset(params: AddAssetParams) {
  return client.post('/user/assets', params) as Promise<AssetData>
}

export function getAsset(id: string) {
  return client.get(`/assets/${encodeURIComponent(id)}`) as Promise<AssetData>
}

export type UpdateAssetParams = AddAssetParams

export function updateAsset(id: string, params: UpdateAssetParams) {
  return client.patch(`/assets/${encodeURIComponent(id)}`, params) as Promise<AssetData>
}

export function deleteAsset(id: string) {
  return client.delete(`/assets/${encodeURIComponent(id)}`) as Promise<void>
}
